From c9bcb26f30a1b2e3434d851aea19200a6d757cca Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Sat, 3 Jun 2023 18:33:54 +0530
Subject: [PATCH] ath11k: fix VLC streaming not working for wan to wlan when
 multicast to unicast flag enable

Currently, additional two bytes after 802.11 header are seen on the air.
When per-packet encap mode is native wifi and vdev level encap mode is
eth mode, native wifi encap functionality is skipped.

This leaves QoS header unremoved in the 802.11 header of the packet. When
sending the packet HW will also add (one more) QoS header. The additional
two bytes before LLC/SNAP header confuses the receiver and those packets
will get dropped at receiver.

Fix this issue by removing QoS header (native wifi encap functionality)
for all the packets which will be enqueued to hw in native wifi mode.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/dp_tx.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath11k/dp_tx.c b/drivers/net/wireless/ath/ath11k/dp_tx.c
index 3d8417d..5c034e6 100644
--- a/drivers/net/wireless/ath/ath11k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_tx.c
@@ -347,7 +347,8 @@ tcl_ring_sel:
 
 	switch (ti.encap_type) {
 	case HAL_TCL_ENCAP_TYPE_NATIVE_WIFI:
-		if (arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED)
+		if ((arvif->vif->offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED) &&
+		     skb->protocol == cpu_to_be16(ETH_P_PAE))
 			is_diff_encap = true;
 		else
 			ath11k_dp_tx_encap_nwifi(skb);
@@ -375,7 +376,7 @@ tcl_ring_sel:
 	if ((!test_bit(ATH11K_FLAG_HW_CRYPTO_DISABLED, &ar->ab->dev_flags) &&
 	    !(info->control.flags & IEEE80211_TX_CTL_HW_80211_ENCAP) &&
 	    !info->control.hw_key && ieee80211_has_protected(hdr->frame_control)) ||
-	    (skb->protocol == cpu_to_be16(ETH_P_PAE) && is_diff_encap)) {
+	    is_diff_encap) {
 		/* HW requirement is that metadata should always point to a
 		 * 8-byte aligned address. So we add alignment pad to start of
 		 * buffer. HTT Metadata should be ensured to be multiple of 8-bytes
-- 
2.17.1

